### DISCUSSION TITLE
A question about the calculation of the number of cpu

### DISCUSSION THREAD

**Discussion by beita1** [tinylibs/tinypool#33] [General]
Hello, I was reading tinypool's source code and had two questions in the `src/physicalCpuCount.ts`.  I don't understand the logic for counting cpu number , could you help me with it?

# questions
- question 1:    Why not just use `os.cpus().length` to get cpu count instead of doing special logic for different plaforms？
- question 2:    Why does the Linux platform get the physical cpu cores by the command `cat /proc/cpuinfo | grep "physical id" | sort | wc -l
` instead of the logical cpu's number by the  command  `cat /proc/cpuinfo| grep "processor"| wc -l`?

**Comment by Aslemammad** [comment]
Hey, Thank you for asking the question! Sure! for the first question, the answer is in the second one! it's because we want to spawn workers based on physical cores, not logical ones!
And why physical cores? Read [this](https://github.com/tinylibs/tinypool/issues/18)

**Reply by beita1** [reply]
Oh, thanks you answer. But I am also confused. 
Based on that issue is to say that getting the physical core count of the CPU is the best way.But The number of physical cores should be equal to the physical CPU multiplied by the number of CPU cores. So the most accurate would be to get the physical CPU by the command `cat /proc/cpuinfo | grep "physical id" | sort | wc -l` and then multiply by the result of `cat /proc/cpuinfo| grep "cpu cores"| uniq`
' to get the number of CPU cores

**Reply by Aslemammad** [reply]
I don't think we need that, because by doing so, we'll not get the physical cores! we should count the physical cores by physical ids AFAIK!

**Reply by beita1** [reply]
I think your understanding of the physical cores is wrong.  Physical cores == physical CPU count *  physical cores.  We can only count the CPU by physical ids. But 1 CPU can have multi-cores.  Such as the following pictures,  2  means the count of physical CPU, and 10 means the physical CPU cores per CPU. So, 20 means physical cores and  40 means logic CPU cores. If you want to get physical cores by physical id, it is not right. 

![image](https://user-images.githubusercontent.com/13329945/177147941-6bea211f-4ae6-4f54-9613-300bbfe668ec.png)

**Reply by Aslemammad** [reply]
I'm open to changing it but that's what I found when reading source codes that wanted to implement physical cores! do you have any article or source code that addresses that?

**Reply by beita1** [reply]
These two articles may be helpful: 
- https://www.howtogeek.com/194756/cpu-basics-multiple-cpus-cores-and-hyper-threading-explained/ 
- https://superuser.com/questions/740611/what-is-the-difference-between-multithreading-and-hyperthreading

And I think the physical core can be calculated like this 
```
    const output = exec('cat /proc/cpuinfo | grep "physical id" | sort | wc -l'); 
    const cpuNum = parseInt(output.trim(), 10);

    const output2 = exec('cat /proc/cpuinfo| grep "cpu cores"| uniq');
    const coreNum = parseInt(output2.trim(), 10);
    const count = cpuNum * coreNum;
```

== edit
something is wrong
```
    const output = exec('cat /proc/cpuinfo | grep "physical id" | sort |uniq | wc -l'); 
    const cpuNum = parseInt(output.trim(), 10);

    const output2 = exec('cat /proc/cpuinfo| grep "cpu cores"| uniq');
    const coreNum = parseInt(output2.trim(), 10);
    const count = cpuNum * coreNum;
```

**Reply by Aslemammad** [reply]
I get it now! That's reasonable! one thing, can we count the coreNum with 
```
cat /proc/cpuinfo | grep "core id" | uniq | wc -l
```
let me know if you think this works for (physical) cores!

**Reply by beita1** [reply]
- It can alos work:  realCpuCount = `cat /proc/cpuinfo | grep "physical id" | sort |uniq | wc -l` * `cat /proc/cpuinfo | grep "core id" | sort | uniq | wc -l`
`
- `cat /proc/cpuinfo | grep "core id" | uniq | wc -l` can't work because core id may be same in the different physical id. For example: 
![image](https://user-images.githubusercontent.com/13329945/177245338-b350b12c-cb4b-45b7-8835-fba52a1fdca3.png)

- Otherwise, we can get physical cores by  `lscpu`: https://unix.stackexchange.com/questions/468766/understanding-output-of-lscpu

**Reply by Aslemammad** [reply]
That's great, could you please try sending a PR?

**Reply by beita1** [reply]
With pleasure, I will send a PR tonight

== edit

This pr has been created: https://github.com/tinylibs/tinypool/pull/34
